yt/yt/server/lib/chunk_pools/ordered_chunk_pool.cpp

```diff
@@ -512,16 +517,27 @@ class TOrderedChunkPool
                 // NB: We use splitting here in order to get the exact same data weight we would get later on in the process.
                 JobCarryOverDataWeight_ = RowCountUntilJobSplit_ ? -chunkSlice->SplitByRowIndex(RowCountUntilJobSplit_).first->GetDataWeight() : 0;
 
-                // If batch row count is zero, there should never be any carry-over data weight,
-                // so the second value cannot be zero for non-empty slices.
-                YT_VERIFY(batchRowCount || RowCountUntilJobSplit_ > 0);
-
                 if (batchRowCount) {
                     YT_VERIFY(*batchRowCount > 0);
                     // Zero rows until split force us to look for the next acceptable split (even when batchRowCountRemainder = 0) since it usually means that a job just ended.
                     if (auto batchRowCountRemainder = (CurrentJob()->GetRowCount() + RowCountUntilJobSplit_) % *batchRowCount; !RowCountUntilJobSplit_ || batchRowCountRemainder) {
                         RowCountUntilJobSplit_ += *batchRowCount - batchRowCountRemainder;
                     }
+                } else if (RowCountUntilJobSplit_ == 0) {
+                    // We should not actually end up here, since both slice addition implementations finish jobs
+                    // as soon as the limit is hit when batch row count is not set.
+                    // However, if we do, let's reset the carry-over data weight, end the job and reprocess the current slice.
+                    YT_LOG_WARNING(
+                        "Creating ordered job prematurely (CurrentJobDataWeight: %v, JobCarryOverDataWeight: %v, ActiveSliceDataWeight: %v, ActiveSliceRowCount: %v)",
+                        GetCurrentJobDataWeight(),
+                        JobCarryOverDataWeight_,
+                        chunkSlice->GetDataWeight(),
+                        chunkSlice->GetRowCount());
+                    JobCarryOverDataWeight_ = 0;
+                    EndJob();
+                    // Current job data weight should be zero now.
+                    pocket = chunkSlice;
+                    continue;
```

Comment:
Looks like `++chunkSliceIndex` is missing.
If this is indeed the case, it might be better to increase it only where it is used
```
auto chunkSlice = pocket ? pocket : chunkSlices[chunkSliceIndex++];
```
Instead of incrementing it on different paths of the loop iteration.